[WEB-3926] Simplify and update Left and right nav sidebar

jamiehenson opened the pull request:
## Description

The start of the big push for new navigation on `docs`. To summarise, it rips almost everything out regarding the current left and right sidebar navigation, and adds in new components to represent the simplified design approach.

Epic: https://ably.atlassian.net/browse/WEB-4082
Left sidebar work: https://ably.atlassian.net/browse/WEB-3926
Right sidebar work: https://ably.atlassian.net/browse/WEB-4090

![Screenshot 2024-12-16 at 17 26 18](https://github.com/user-attachments/assets/4d9a5b89-9d77-4f4d-b9c5-ab71776bc516)


## Review

Instructions on how to review the PR. 

- [Heroku](https://ably-docs-next-staging.herokuapp.com/docs/)


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced a new environment variable `INKEEP_CHAT_ORGANIZATION_ID` in the example environment configuration.
	- Added a streamlined breadcrumb level calculation and enhanced inline Table of Contents processing.
	- New default export structure for navigation and language data across various product categories.
	- New `languageLabel` function to format language labels and versions.
	- New `Breadcrumbs` component for navigation tracking.
	- New `LanguageSelector` component for selecting programming languages.

- **Bug Fixes**
	- Adjusted header margins for better spacing in the `BodySection` component.

- **Documentation**
	- Comprehensive tests added for `LanguageSelector` and `Breadcrumbs` components.

- **Refactor**
	- Simplified layout structure across various pages by removing unnecessary sidebar context.
	- Updated various components to improve type safety and streamline props.

- **Chores**
	- Removed several unused components and files related to sidebar functionality.

- **Style**
	- Adjusted CSS styles for dropdown and navigation components for improved layout and responsiveness.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

jamiehenson commented:
@coderabbitai review

jamiehenson commented on src/data/nav/livesync.ts:
```diff
@@ -0,0 +1,48 @@
+    {
+      name: 'API References',
+      pages: [
+        {
+          link: 'https://sdk.ably.com/builds/ably-labs/models/main/typedoc/index.html',
+          name: 'SDK',
+          external: true,
+        },
```
Not needed, we're not defining all link props here

jamiehenson commented on src/data/nav/platform.ts:
```diff
@@ -0,0 +1,131 @@
+      ],
+    },
+    {
+      name: 'Pricing',
+      pages: [
+        {
+          link: '/pricing',
+          name: 'Pricing',
+        },
+        {
+          name: 'Hello',
+          pages: [
+            {
+              link: '/pricing/standard',
+              name: 'Pricing Standard',
+            },
+            {
+              link: '/pricing/pro',
+              name: 'Pricing Pro',
+            },
+            {
+              link: '/pricing/enterprise',
+              name: 'Pricing Enterprise',
+            },
+          ],
+        },
+        {
+          link: '/pricing/billing',
+          name: 'Pricing Billing',
+        },
+        {
+          link: '/pricing/limits',
+          name: 'Pricing Limits',
+        },
+        {
+          link: '/pricing/faqs',
+          name: 'Pricing Faqs',
+        },
+      ],
+    },
```
This is test data, duplicates will be ironed out later when accurate date is provided

jamiehenson commented:
@coderabbitai review

jamiehenson commented on src/data/nav/types.ts:
```diff
@@ -0,0 +1,25 @@
+import { IconName } from '@ably/ui/core/Icon/types';
+
+export type NavProduct = {
+  name: string;
+  icon: { closed: IconName; open: IconName };
+  content: NavProductContent[];
+  api: NavProductContent[];
+  showJumpLink?: boolean;
+  link?: string;
+};
```
We do want the "Content -> Pages -> Content -> Pages" in this scenario as we want to support an arbitrary number of layers deep in the nav hierarchy. How does aliasing `NavProductSection` solve this?

jamiehenson replied:
Thanks bunno

jamiehenson commented on src/data/nav/livesync.ts:
```diff
@@ -0,0 +1,49 @@
+
+export default {
+  name: 'Ably LiveSync',
+  link: '/products/livesync',
+  icon: {
```
Don't worry about it, this is incomplete test data.

jamiehenson commented on src/components/Layout/LanguageDropdown.tsx:
```diff
@@ -0,0 +1,128 @@
+const LanguageDropdownOption = ({ isOption, setMenuOpen, langParam, ...props }: LanguageDropdownOptionProps) => (
+  <div
+    onClick={() => setMenuOpen(!props.selectProps.menuIsOpen)}
+    className={cn('group/lang-dropdown flex gap-8 items-center rounded', {
+      'p-8 hover:bg-neutral-100 dark:hover:bg-neutral-1200 mx-8 last:mb-8': isOption,
+    })}
```
How it is currently is a valid application of the `clsx` library

jamiehenson commented on src/components/Layout/RightSidebar.tsx:
```diff
@@ -0,0 +1,154 @@
+      typeof document !== 'undefined' ? document.querySelector('article')?.querySelectorAll('h2, h3') ?? [] : [];
+    const headerData = Array.from(headerElements)
+      .filter((element) => element.id && element.textContent)
+      .map((header) => ({
+        type: header.tagName,
+        label: header.textContent ?? '',
+        id: header.id,
+        height: header.getBoundingClientRect().height,
+      }));
+
+    setHeaders(headerData);
+
+    const handleIntersect = (
+      entries: {
+        target: Element;
+        isIntersecting: boolean;
+      }[],
+    ) => {
+      entries.forEach((entry) => {
+        if (entry.isIntersecting && entry.target.id) {
+          setActiveHeader({
+            id: entry.target.id,
+          });
+        }
+      });
+    };
+
+    observer.current = new IntersectionObserver(handleIntersect, {
+      root: null,
+      threshold: 1,
+    });
+
+    headerElements.forEach((header) => {
+      observer.current?.observe(header);
+    });
+
+    return () => {
+      observer.current?.disconnect();
+    };
+  }, []);
```
This doesn't solve anything. The original definition of `headerElements` has an empty array if no elements are found, which allows for a graceful bailout of any further logic down the line.

jamiehenson commented on src/contexts/layout-context.tsx:
```diff
@@ -0,0 +1,71 @@
+  );
+};
+
+export const useLayoutContext = () => {
+  return useContext(LayoutContext);
+};
```
Sure why not

jamiehenson commented on src/pages/how-to/pub-sub.tsx:
```diff
@@ -127,71 +126,69 @@ const PubSubHowTo = () => {
+                      <SandpackPreview
+                        style={{ height: '480px' }}
+                        className="rounded-lg overflow-hidden"
+                        showOpenInCodeSandbox={false}
+                        showRefreshButton
                       />
-
-                      <div className="flex gap-16 my-16 flex-col sm:flex-row">
-                        <SandpackPreview
-                          style={{ height: '480px' }}
-                          className="rounded-lg overflow-hidden"
-                          showOpenInCodeSandbox={false}
-                          showRefreshButton
-                        />
-                        <SandpackPreview
-                          style={{ height: '480px' }}
-                          className="rounded-lg overflow-hidden"
-                          showOpenInCodeSandbox={false}
-                          startRoute="/?publisher=false"
-                        />
-                      </div>
-                    </SandpackProvider>
-                  </>
-                ) : (
-                  <b>Loading...</b>
-                )}
-              </div>
-            </aside>
-          </article>
-        </Layout>
-      </SidebarProvider>
+                      <SandpackPreview
+                        style={{ height: '480px' }}
+                        className="rounded-lg overflow-hidden"
+                        showOpenInCodeSandbox={false}
+                        startRoute="/?publisher=false"
+                      />
+                    </div>
+                  </SandpackProvider>
+                </>
```
Editing this page beyond the minimum for navigation support is out of scope for the core goals of this PR.

jamiehenson commented:
@coderabbitai review

kennethkalmer commented on src/components/blocks/software/__snapshots__/Pre.test.tsx.snap:
```diff
@@ -320,7 +320,7 @@ exports[`<Pre /> should successfully render code elements with  only Rest langua
                   class="ably-select__single-value css-9aszhm-singleValue"
                 >
-                  JavaScript v2.0
+                  rest_javascript
```
Uhm, are these snapshots revealing something here...?

jamiehenson replied:
This is a false alarm I think, the language dropdown definitely doesn't show these ids instead of full labels - this'll very likely be around how the test data is set up for the dropdown component

kennethkalmer commented on src/data/languages/languageData.ts:
```diff
@@ -0,0 +1,32 @@
+export default {
+  platform: {} as LanguageData,
+  pubsub: {
+    javascript: 1.2,
```
Oh right, this might be why I kept being confused seeing JS 1.2 in the snapshots... Is this right though? I was certain we're on 2.x for JS

jamiehenson replied:
This data is inaccurate atm, most of it is bot-vomited. Will need some fine tuning in the QA period with Mark etc but the key thing is that the data structures provide the framework we need

kennethkalmer commented on src/data/languages/types.ts:
```diff
@@ -0,0 +1,30 @@
+const languageKeys = [
```
[question] Could this be derived from `languageInfo.ts`? Then we have a single source of truth :)

jamiehenson replied:
As it stands this would be a circular dependency as the object in `languageInfo` is typed from this - but tbh that's overtyped, we can just do as you suggest

kennethkalmer reviewed (commented):
First sweep of the code with some questions, have to play with the app a bit to give UI/UX feedback.

This is an amazing amount of work @jamiehenson, thank you so much!

jamiehenson commented on src/components/LanguageButton/LanguageButton.test.tsx:
```diff
@@ -76,9 +76,7 @@ describe(`<LanguageButton />`, () => {
-      >
-        Java
-      </button>
+      />
```
There's something iffy with retrofitting these old language buttons as it's smooshing together two different conventions of defining languages, I considered phasing these out as well but this is the boundary of scope I put the flag down on. I'll take another look at those tomorrow

aralovelace commented on src/components/Layout/LanguageSelector.tsx:
```diff
@@ -0,0 +1,131 @@
+import cn from '@ably/ui/core/utils/cn';
+import { useRef, useState } from 'react';
+import Select from 'react-select';
+import { languageData } from 'src/data/languages';
```
we might able to save 1 line by combining the two imports?

`import { languageData, languageInfo } from 'src/data/languages';`

aralovelace commented on src/components/Layout/Layout.tsx:
```diff
@@ -1,106 +1,42 @@
-import cn from '@ably/ui/core/utils/cn';
-import { ReactNode, useEffect } from 'react';
+import { PropsWithChildren } from 'react';
```
might need to import `React` since we use the variable?

aralovelace commented on src/components/Layout/utils.ts:
```diff
@@ -0,0 +1,121 @@
+      If the page matches a product index link, return details for the product
+    */
+    if (data[key].nav.link === strippedTargetLink) {
+      const { name, link } = data[key].nav;
```
the link might be undefined? shall we add a default string?

`const { name, link = '' } = data[key].nav;`

aralovelace commented on src/pages/index.tsx:
```diff
@@ -5,8 +5,6 @@ import Layout from 'src/components/Layout';
 import { HomepageContent, SectionProps } from 'src/components/Homepage/HomepageContent';
-
-import { SidebarProvider } from 'src/contexts/SidebarContext';
 import { PageLanguageProvider } from 'src/contexts';
```
the `ABLY_MAIN_WEBSITE` might need to remove as its not being used?

aralovelace reviewed (changes requested):
Looks so nice! A lot nicer!
Only few comments and found some visual issue:

1- If the page is too long the GIthub buttons on the right overflows
![image](https://github.com/user-attachments/assets/5cd8d512-5920-44c7-85de-b4940190f03d)

2- No link yet for the Request changes button?

3- When the RHS is too long it look like this: `api/rest-api`

![image](https://github.com/user-attachments/assets/25a4f307-e94e-4ec4-a5ac-0cccbd48603b)

4- Mobile view on this page `/spaces/space#update-profile`

<img width="429" alt="image" src="https://github.com/user-attachments/assets/17731770-bd91-4297-b7d6-26635c242575" />

That's all

jamiehenson commented:
@coderabbitai review

jamiehenson commented:
@coderabbitai review

jamiehenson commented on src/data/nav/platform.ts:
```diff
@@ -0,0 +1,131 @@
+    },
+    {
+      name: 'Pricing',
+      pages: [
+        {
+          link: '/pricing',
+          name: 'Pricing',
+        },
+        {
+          name: 'Hello',
+          pages: [
+            {
+              link: '/pricing/standard',
+              name: 'Pricing Standard',
+            },
+            {
+              link: '/pricing/pro',
+              name: 'Pricing Pro',
+            },
+            {
+              link: '/pricing/enterprise',
+              name: 'Pricing Enterprise',
+            },
+          ],
+        },
+        {
+          link: '/pricing/billing',
+          name: 'Pricing Billing',
+        },
+        {
+          link: '/pricing/limits',
+          name: 'Pricing Limits',
+        },
+        {
+          link: '/pricing/faqs',
+          name: 'Pricing Faqs',
+        },
+      ],
+    },
+  ],
```
Don't suggest changes to src/data/* - the data is currently incomplete and needs to be updated later.

jamiehenson commented on src/components/Layout/RightSidebar.test.tsx:
```diff
@@ -0,0 +1,96 @@
+    const headerLink = screen.getByRole('link', { name: 'Header 1' });
+    fireEvent.click(headerLink);
+    expect(headerLink).toHaveClass('font-bold');
+    cleanup();
```
🤖 

I'll move it to afterEach

matt423 commented:
Haven't given the code a thorough walk through yet but have noticed this behaviour when viewing the site. For menu's that are lower in the nav, the focus isn't quite right. Happens via the menu or direct url, such as [this one](https://ably-docs-next-staging.herokuapp.com/docs/api/realtime-sdk)

[Screencast from 2025-01-07 11-06-31.webm](https://github.com/user-attachments/assets/cf336b9b-dce9-43bc-9916-a6b33d8df8cb)

jamiehenson commented on src/components/Layout/RightSidebar.tsx:
```diff
@@ -0,0 +1,153 @@
+  {
+    label: 'Edit on GitHub',
+    icon: 'icon-social-github',
+    link: `${githubEditPath}${location.pathname}.textile`,
```
We can fix this logic by scrubbing off the `docs/` from the path and adding something like an `indexPage` field to the nav data to determine whether we access `GITHUB/thing/index.textile` or `GITHUB/thing.textile`. I'm likely missing something but I don't see how higher queries would provide more accurate context to link to Github directories here.

Also, what would be affected by the MDX project, would it be a different folder structure that we could no longer map to? I had imagined that the existing textile files would just be replaced in situ.

jamiehenson commented:
The bot's summary of the latest changes, my annotations in brackets:

Types:
- Added an optional indexPage property to the NavProductPage type (assists with the Edit on Github link population).

Navigation Data:
- Updated navigation data to include the indexPage property for the 'Channels' link (more data needs to be added for other pages later).

Layout Utilities:
- Enhanced the determineActivePage function to handle the page object more comprehensively (adds all page object information to the returned data instead of just name and link).
- Added a new utility function hierarchicalKey for generating hierarchical keys (cleans up duplicate key errors with identical links by including the page hierarchy in the component key).

Right Sidebar:
- Modified external links to dynamically generate GitHub edit and request paths based on the current page.
- Included language-specific query parameters in sidebar links (sorta, languages are persisted across page changes now).
- Updated styles and hover effects for sidebar links.

Left Sidebar:
- Introduced hierarchical keys for better link handling.
- Removed the useEffect hook that scrolls to the selected link on mount (took out the scroll-to-link-on-load functionality as it was inaccurate as Matt points out above and annoying as Mark pointed out).

Breadcrumbs:
- Utilized the new hierarchicalKey to generate keys for breadcrumb links.

Tests:
- Added cleanup to afterEach in RightSidebar tests.
- Removed test related to scrolling to the selected link on mount in LeftSidebar tests.

kennethkalmer reviewed (approved):
![](https://i.giphy.com/media/v1.Y2lkPTc5MGI3NjExdXJqZzN6ZnFjczAwdG03Z3p2NGZkYXptNzQ0dHdldWttd3B3cDB3cCZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/5fORBo8Sfcs5bfnc6N/giphy.gif)
